diff --git a/tccrun.c b/tccrun.c
old mode 100644
new mode 100755
index 79dfd08..eb779f1
--- a/tccrun.c
+++ b/tccrun.c
@@ -115,7 +115,10 @@ static int rt_mem(TCCState *s1, int size)
 #ifdef CONFIG_SELINUX
     /* Using mmap instead of malloc */
     void *prw;
-    char tmpfname[] = "/tmp/.tccrunXXXXXX";
+    const char *tmpdir = getenv("TMPDIR");
+    if (!tmpdir || !*tmpdir) tmpdir = "/tmp";
+    char tmpfname[256];
+    snprintf(tmpfname, sizeof(tmpfname), "%s/.tccrunXXXXXX", tmpdir);
     int fd = mkstemp(tmpfname);
     unlink(tmpfname);
     ftruncate(fd, size);
